### PR TITLE
Fix possible file corruption during backup or power loss or other abn…

### LINKED ISSUE #240: Find/Replace On Large File Causes Crash
When I try to run a regular-expression find/replace operation on a large file, Notepad++ crashes.
- I don't think it's the fault of the regex. It's relatively simple - no backreferences, no nested repeats, and no repeatable sections which can match a newline (so it can't go mad and try to match the entire file)
- The thing I am replacing with is the blank string, so I'm pretty sure that's not the problem.
- It is definitely correlated with the size of the input to the find/replace; if I select a small enough part of the file and use 'In Selection', no crash:
  - Selection: 422932 characters, 8816 lines. Result: 1759 matches found and replaced, no crash
  - Full file: 1555727characters, 32212 lines. Result: 6433 matches found and replaced, but Notepad++ crashes as soon as the find/replace operation finishes.
- The crash appears to take place after the find/replace operation has finished - the 'X occurrences replaced' status on the Replace dialogue updates, and in the background I can see the file being re-rendered - but before control is actually returned to me after I have clicked the 'Replace All' button
- Error messages are as follows:
  - Win32Exception: "An exception occurred. Notepad++ cannot recover and must be shut down.
    The exception details are as follows:
    Code: 0xC0000005
    Type: Access violation
    Exception address: 0x5DCB0F3A
  - Do you want to save a dump file?
    Doing so can aid in developing
    [Yes] [No](I select yes.)
  - Failed to create dump file 'C:\Program Files (x86)\Notepad++\NppDump.dmp' (error 5)
  - Notepad++ will attempt to save any unsaved data. However, dataloss is very likely.
  - Notepad++ was able to successfully recover some unsaved documents, or nothing to be saved could be found.
    You can find the results at:
    C:\Users[redacted]\AppData\Local\Temp\N++RECOV
  - Notepad++: a free (GNU source code editor has stopped working
    A problem has caused the program to stop working correctly.
    Windows will close the program and notify you if a solution is available.
  - Sometimes an additional dialogue complaining about a 'pure virtual function call' appears, but I haven't been able to reproduce that reliably so I don't have the details.
- It does not appear to be correlated with memory usage; I watched the task manager during the find/replace and crash, and notepad++.exe's memory usage remained level throughout
- It appears to be correlated with the number (or possibly total size) of matches; I changed my regex by one character so it would require the same search pattern but fail to find any matches, and I got no crash.

### PR DESCRIPTION
**Pull request by pnedev**
…ormal N++ termination

Fixes #240, #2381, #2883, #4346, #4655 and probably more issues related to loss of data.

### CODE REVIEW

**Comment by pnedev** [comment]
Due to the synchronization issues with backup operation I've made the backup run in the main N++ thread (the UI one) by posting in its message queue. This should cause any current long running user operation to first finish and then execute the backup save.

The problems so far were caused by the fact that some complex user operations were not considered _'long-running'_ (but actually are on very big files) and were not protected by the `LongRunningOperation` lock.

Strictly speaking, the `LongRunningOperation` lock during backup is not needed anymore because now all Scintilla buffer operations are serialized via channeling them through a single thread but I have left them in case I have overlooked something.

**Comment by Barbapoux** [comment]
Hi, just to confirm case #240 with last hypothesis of @ClaudiaFrank 
when instant backup period is greater than duration for replacement, the crash doesn't occur anymore.
Thanks

**Comment by donho** [comment]
@Barbapoux Can you reproduce steadily this issue?

**Comment by Barbapoux** [comment]
Oui tout à fait. Je pense que tout le monde peut le faire.
Si ta question est "Est-ce que tu peux tester la version corrigée?", j'ai de nouveaux fichiers (125 Mo), donc oui c'est possible. Il faut juste m'indiquer la procédure souhaitée.
Cordialement.

**Comment by donho** [comment]
@Barbapoux 
Cool ! Pourrais-tu me donner les instructions pour que je puisse le reproduire stp?

**Comment by donho** [comment]
@pnedev Could you solve the conflicts please?

**Comment by pnedev** [comment]
@donho Done.

**Comment by Barbapoux** [comment]
Pardon, tout est décrit dans le thread #240 (https://github.com/notepad-plus-plus/notepad-plus-plus/issues/240#issuecomment-200867146) , il y a un fichier pour le test, il suffit de l'éditer et faire supprimer les blancs de fin de ligne (ou toute opération un peu longue de remplacement, qui va mettre le bazar dans les pointeurs).
Avec la sauvegarde instantané à 7 secondes par défaut, en général, ça va au tas.

**Comment by donho** [comment]
@Barbapoux  This PR is about fixing possible file corruption during backup. 
Whereas your comment is about Crash for Find/Replace on Large File.
I don't see the relation between these 2 issues.

**Comment by Barbapoux** [comment]
I think @pnedev is on the right way. I had some issues with long replacement operations (replace all xxx xxx xxx by nothing on big file) after version 6.5.5, without any plugin, and I have found that issue with trailing blank that does the same error and easily reproductible with dummy data. On my point of view, the reason was the same, it seems you have solved replacement crash ( it was 2 years ago), as I don't have anymore garbage in my file with current version (replacement generated crashs or memory dump in middle of text file).
As @ClaudiaFrank has seen, disabling the periodic backup is solving the issue on trailing blank, then it is certainly because it is trying to copy a memory block that has changed.

File: PowerEditor/src/ScitillaComponent/Buffer.cpp [hunk 1]
```diff
@@ -957,47 +922,8 @@ bool FileManager::backupCurrentBuffer()
...
-			printStr(TEXT("WaitForSingleObject problem in deleteBufferBackup()!"));
-			return false;
-		}
-
-		// unlocled here, set to non-signaled state, to block all threads
-		::ResetEvent(writeEvent);
-	}
-
-	EventReset reset(writeEvent); // Will reset event in destructor.
-
```
Comment (donho) [inline PowerEditor/src/ScitillaComponent/Buffer.cpp]: No `LongRunningOperation op;` instruction here?

Comment (pnedev) [reply PowerEditor/src/ScitillaComponent/Buffer.cpp]: You mean in `FileManager::deleteBufferBackup()` ?
It won't hurt to add `LongRunningOperation op` but strictly speaking it is not needed at all. All buffer operations are now channeled through a single thread so no need for locking at all.
I can either remove all `LongRunningOperation` locks from `FileManager` in this file or add a lock in `deleteBufferBackup()` for consistency. What do you prefer?

Comment (donho) [reply PowerEditor/src/ScitillaComponent/Buffer.cpp]: > All buffer operations are now channeled through a single thread

Could you point me this part?

Comment (pnedev) [reply PowerEditor/src/ScitillaComponent/Buffer.cpp]: Yes, at the very beginning of this Git diff I have changed the direct call to `MainFileManager->backupCurrentBuffer()` to `::PostMessage(Notepad_plus_Window::gNppHWND, NPPM_INTERNAL_SAVEBACKUP, 0, 0);`.
This way the actual buffer backup will go through the message queue which is serialized - no concurrency.
The other operations here were not synchronized (locked) anyway but they also are always triggered from the GUI thread (again through the message queue). Am I missing anything?

Comment (donho) [reply PowerEditor/src/ScitillaComponent/Buffer.cpp]: Well played!

**Pull request closed without merging**